lib/dw/utils/parser.test.js

```diff
@@ -0,0 +1,234 @@
+    t.is(expr.evaluate({ name: 'Datawrapper', start: 3, len: 6 }), 'awrapp');
+});
+
+test('concat function', t => {
+    const expr = parser.parse('CONCAT(first, " ", last)');
+    t.is(expr.evaluate({ first: 'Lorem', last: 'Ipsum' }), 'Lorem Ipsum');
+});
+
+test('trim function', t => {
+    const expr = parser.parse('TRIM(name)');
+    t.is(expr.evaluate({ name: '  spaces\t\t' }), 'spaces');
+});
+
+test('trim function without brackets', t => {
+    const expr = parser.parse('TRIM name');
+    t.is(expr.evaluate({ name: '  spaces\t\t' }), 'spaces');
+});
+
+test('TRUE', t => {
+    const expr = parser.parse('TRUE');
+    t.is(expr.evaluate({ name: '' }), true);
+});
+
+test('NOT TRUE', t => {
+    const expr = parser.parse('NOT TRUE');
+    t.is(expr.evaluate({ name: '' }), false);
+});
+
+test('IF(TRUE', t => {
+    const expr = parser.parse('IF(TRUE, "yes", "no")');
+    t.is(expr.evaluate({ name: '' }), 'yes');
+});
+
+test('FALSE', t => {
+    const expr = parser.parse('FALSE');
+    t.is(expr.evaluate({ name: '' }), false);
+});
+
+test('TRUE and FALSE', t => {
+    const expr = parser.parse('TRUE and FALSE');
+    t.is(expr.evaluate({ name: '' }), false);
+});
+
+test('true', t => {
+    const expr = parser.parse('true');
+    t.is(expr.evaluate({ true: 12 }), 12);
+    t.throws(() => expr.evaluate({ foo: 12 }));
+});
+
+test('ABS()', t => {
+    const expr = parser.parse('ABS(a)');
+    t.is(expr.evaluate({ a: -12 }), 12);
+    t.is(expr.evaluate({ a: 10 }), 10);
+});
+
+test('ABS', t => {
+    const expr = parser.parse('ABS a');
+    t.is(expr.evaluate({ a: -12 }), 12);
+    t.is(expr.evaluate({ a: 10 }), 10);
+});
+
+test('ROUND()', t => {
+    const expr = parser.parse('ROUND(a)');
+    t.is(expr.evaluate({ a: -12.345 }), -12);
+    t.is(expr.evaluate({ a: 10.56 }), 11);
+});
+
+test('ROUND(a,1)', t => {
+    const expr = parser.parse('ROUND(a, 1)');
+    t.is(expr.evaluate({ a: -12.345 }), -12.3);
+    t.is(expr.evaluate({ a: 10.56 }), 10.6);
+});
+
+test('year() function', t => {
+    const expr = parser.parse('YEAR(date)');
+    t.is(expr.evaluate({ date: new Date(2020, 1, 1) }), 2020);
+    t.is(expr.evaluate({ date: new Date(2017, 4, 14) }), 2017);
+    t.is(expr.evaluate({ date: '2017-06-23' }), 2017);
+});
+
+test('month() function', t => {
+    const expr = parser.parse('MONTH(date)');
+    t.is(expr.evaluate({ date: new Date(2020, 1, 1) }), 2);
+    t.is(expr.evaluate({ date: new Date(2017, 4, 14) }), 5);
+});
+
+test('day() function', t => {
+    const expr = parser.parse('DAY(date)');
+    t.is(expr.evaluate({ date: new Date(2020, 1, 1) }), 1);
+    t.is(expr.evaluate({ date: new Date(2017, 4, 14) }), 14);
+});
+
+test('hours() function', t => {
+    const expr = parser.parse('HOURS(date)');
+    t.is(expr.evaluate({ date: new Date(2020, 1, 1) }), 0);
+    t.is(expr.evaluate({ date: new Date(2017, 4, 14, 18, 30, 5) }), 18);
+});
+
+test('sin a^b', t => {
```

Comment:
Would the parser understand `a ^ b` too?

Reply:
sure